// storybook-directory-chapters/lib/loadDirectories.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = loadDirectories;

var _react = require('@storybook/react');

var initData = { dirs: [] };

function loadDirectories(context) {
  context.keys().reduce(function (acc, c) {
    var dirs = getDirs(c);

    if (!dirs.length) return acc;

    var stories = void 0;
    var lastDirs = acc.dirs.concat();

    if (acc.stories) {
      stories = Object.assign({}, acc.stories);
    } else {
      stories = createStories(dirs[0]);
      lastDirs = [dirs[0]];
    }

    // check the diff between the current path and the previous path
    var cursorsToMove = cursorsToPath(lastDirs, dirs);

    // check how many chapters to end
    stories = endChapters(stories, cursorsToMove.decrease);

    // if top level directories change, create new stories and reset the cursors
    if (shouldCreateStories(lastDirs, dirs)) {
      stories = createStories(dirs[0]);
      lastDirs = [dirs[0]];
      cursorsToMove = cursorsToPath(lastDirs, dirs);
    }

    // check how many chapters to add
    stories = addChapters(stories, cursorsToMove.increase, dirs.slice(lastDirs.length - cursorsToMove.decrease));

    stories = context(c)(stories);
    return { dirs: dirs, stories: stories };
  }, initData);
}

function getDirs(path) {
  return path.replace(/..?\//, '').split('/').reverse().slice(1).reverse();
}

function shouldCreateStories(a, b) {
  return a[0] !== b[0];
}

function createStories(name) {
  return (0, _react.storiesOf)(name, module);
}

function addChapters(stories, i, dirs) {
  var name = void 0;
  switch (i) {
    case 0:
      return stories;
    default:
      name = dirs.concat().reverse()[i - 1];
      return addChapters(stories.chapter(name), i - 1, dirs);
  }
}

function endChapters(stories, i) {
  switch (i) {
    case 0:
      return stories;
    default:
      return endChapters(stories.endOfChapter(), i - 1);
  }
}

function cursorsToPath(a, b) {

  var dt = a.length - b.length;
  var min = Math.min(a.length, b.length);
  var max = Math.max(a.length, b.length);

  var cursors = { increase: 0, decrease: 0 };

  for (var i = 0; i < min; i++) {
    if (a[i] === b[i]) continue;

    if (dt <= 0) {
      // got larger or the same
      // it increases from the diff point even when the lengths are the same
      cursors.increase = max - i;
    }
    // either got larger or the same or got smaller
    cursors.decrease = min - i;
    return cursors;
  }

  if (dt < 0) {
    // got larger
    cursors.increase = max - min;
  } else if (dt > 0) {
    // got smaller
    cursors.decrease = max - min; // (min - max) * -1
  }
  return cursors;
}
