// storybook-adk/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addonManager = exports.register = exports.addonComposer = exports.CSS_CLASS = exports.EVENT_ID_DATA = exports.EVENT_ID_INIT = exports.PANEL_ID = exports.ADDON_TITLE = exports.ADDON_ID = undefined;

var _addonComposer = require('./store/addonComposer');

Object.defineProperty(exports, 'addonComposer', {
    enumerable: true,
    get: function get() {
        return _addonComposer.addonComposer;
    }
});
exports.initStore = initStore;

var _register = require('./register');

var _register2 = _interopRequireDefault(_register);

var _decorator = require('./store/decorator');

var addonManager = _interopRequireWildcard(_decorator);

var _store = require('./store/store');

var _store2 = _interopRequireDefault(_store);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ADDON_ID = exports.ADDON_ID = 'sm/storybook-adk'; // note: addons, panels and events get unique names using a prefix
var ADDON_TITLE = exports.ADDON_TITLE = 'PODDA';
var PANEL_ID = exports.PANEL_ID = ADDON_ID + '/adk-panel';
var EVENT_ID_INIT = exports.EVENT_ID_INIT = ADDON_ID + '/adk-event/init';
var EVENT_ID_DATA = exports.EVENT_ID_DATA = ADDON_ID + '/adk-event/data';
var CSS_CLASS = exports.CSS_CLASS = 'adk-ui';

exports.register = _register2.default;
exports.addonManager = addonManager;
function initStore() {
    return _store2.default.apply(undefined, arguments);
}
