// storybook-adk/dist/register.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

exports.default = function (addonSettings, onRegister) {
    var settings = addonSettings.config;
    //    const { defaultData, addonApi, config } = addonSettings;
    _addons2.default.register(settings.ADDON_ID, function (api) {
        var addonStoreEnv = (0, _store2.default)(addonSettings, api);
        var PanelContainer = (0, _composer2.default)(addonStoreEnv);
        var getID = function getID() {
            return 'pd' + Math.round(Math.random() * 100);
        };
        var addonPanel = addonSettings.render ? addonSettings.render(addonStoreEnv) : null;
        var addonDisabled = addonSettings.renderDisabled ? addonSettings.renderDisabled(addonStoreEnv) : null;
        var addonRender = addonPanel || addonDisabled || null;

        if (addonRender) {
            _addons2.default.addPanel(settings.PANEL_ID, {
                title: settings.ADDON_TITLE,
                render: function render() {
                    return _react2.default.createElement(PanelContainer, {
                        api: api,
                        addonControl: null,
                        initData: addonSettings.initData,
                        rootProps: { enquiry: _store.ENQ_ASK, ID: getID() },
                        addonRender: addonPanel(),
                        addonRenderDisabled: addonDisabled ? addonDisabled() : null,
                        style: { width: '100%' },
                        className: settings.CSS_CLASS + '-panel',
                        onChannelInit: addonSettings.onChannelInit || null
                    });
                }
            });
        }

        if (onRegister) onRegister(addonStoreEnv);
    });
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _addons = require('@storybook/addons');

var _addons2 = _interopRequireDefault(_addons);

var _store = require('./store/store');

var _store2 = _interopRequireDefault(_store);

var _composer = require('./store/composer');

var _composer2 = _interopRequireDefault(_composer);

var _ = require('./');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
