// @storybook/addon-chapters/dist/chapters/defaults.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BADGES = undefined;
exports.Breadcrumbs = Breadcrumbs;
exports.chapterTOC = chapterTOC;
exports.bookmarkList = bookmarkList;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _addonLinks = require('@storybook/addon-links');

var _navigate = require('./navigate');

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BADGES = exports.BADGES = {
  up: 'â¤´ï¸',
  toc: 'â„¹ï¸',
  bookmarks: 'ðŸ”–',
  tags: ':hash:'
};

var styles = {
  toc: {
    fontFamily: 'sans-serif'
  },
  chapList: {
    listStyleType: 'circle'
  },
  chapSection: {
    backgroundColor: '#f7f7f7',
    fontSize: 14
  },
  chapLink: {
    textDecoration: 'none',
    color: 'black'
  },
  chapItems: {
    padding: 6
  },
  storySection: {
    paddingLeft: 16,
    fontSize: 12
  },
  storyLink: {
    textDecoration: 'none',
    color: 'black'
  },
  storyItems: {
    padding: 4
  },
  crumbs: {
    fontSize: 12
  }
};

function Breadcrumbs(_ref) {
  var path = _ref.path;
  var currentStory = _ref.currentStory;

  var steps = path.map(function (chapter) {
    return _react2.default.createElement(
      'span',
      { style: styles.crumbs, key: chapter.name },
      ' ',
      _react2.default.createElement(
        'a',
        {
          href: '#' + chapter.name,
          onClick: (0, _navigate.chapterSelect)(chapter, currentStory)
        },
        chapter.name
      ),
      ' /'
    );
  });
  return _react2.default.createElement(
    'div',
    null,
    steps
  );
}
Breadcrumbs.propTypes = {
  path: _propTypes2.default.arrayOf(_propTypes2.default.object),
  currentStory: _propTypes2.default.string
};

function chapterTOC(chapter) {
  return function () {
    var subchapters = chapter.subchapters.map(function (val) {
      return _react2.default.createElement(
        'li',
        { style: styles.chapItems, key: val.name },
        _react2.default.createElement(
          'a',
          {
            href: '#' + val.name,
            style: styles.chapLink,
            onClick: (0, _addonLinks.linkTo)(chapter.name, '[' + val.name + ']')
          },
          val.name
        )
      );
    });
    var stories = chapter.stories.map(function (val) {
      return _react2.default.createElement(
        'li',
        { style: styles.storyItems, key: val.storyName },
        _react2.default.createElement(
          'a',
          {
            href: '#' + val.storyName,
            style: styles.chapLink,
            onClick: (0, _addonLinks.linkTo)(chapter.name, '' + val.storyName)
          },
          val.storyName
        )
      );
    });

    return _react2.default.createElement(
      'div',
      { style: styles.toc },
      _react2.default.createElement(Breadcrumbs, { path: (0, _utils.breadcrumbs)(chapter), currentStory: chapter.name }),
      _react2.default.createElement(
        'h3',
        null,
        chapter.name
      ),
      _react2.default.createElement('hr', null),
      subchapters.length > 0 ? _react2.default.createElement(
        'div',
        { style: styles.chapSection },
        _react2.default.createElement(
          'ul',
          { style: styles.chapList },
          subchapters
        )
      ) : null,
      _react2.default.createElement(
        'div',
        { style: styles.storySection },
        _react2.default.createElement(
          'ul',
          null,
          stories
        )
      )
    );
  };
}

function bookmarkList(chapter) {
  return function () {
    var bookmarks = chapter.bookmarks.map(function (val) {
      return _react2.default.createElement(
        'li',
        { style: styles.storyItems, key: val.storyName },
        _react2.default.createElement(
          'a',
          {
            href: '#' + val.storyName,
            style: styles.chapLink,
            onClick: function onClick() {
              return console.log(val.storyName);
            }
          },
          val.storyName
        )
      );
    });
    return _react2.default.createElement(
      'div',
      { style: styles.toc },
      _react2.default.createElement(Breadcrumbs, { path: (0, _utils.breadcrumbs)(chapter), currentStory: chapter.name }),
      _react2.default.createElement(
        'h3',
        null,
        chapter.name,
        ' bookmarks:'
      ),
      _react2.default.createElement('hr', null),
      _react2.default.createElement(
        'div',
        { style: styles.storySection },
        _react2.default.createElement(
          'ul',
          null,
          bookmarks
        )
      )
    );
  };
}
